### DISCUSSION TITLE
Type safety for middleware that sets context variables

### DISCUSSION THREAD

**Discussion by pdlug** [honojs/hono#3257] [Q&A]
I have an authentication middleware which ensures the user is logged in and adds the `user` to the context if they are. My middleware adds the type of this context variable per the docs:

```typescript
declare module "hono" {
  interface ContextVariableMap {
    user: User;
  }
}
```

The problem is, all handlers see `c.get('user')` as being present even if they don't use the middleware. If I make `user` optional then in every handler that uses my middleware I need a guard `if (!user) return c.json(...)` which is redundant because the middleware is already guaranteeing that the context variable is set.

Is there any way around this?

**Comment by nmcdaines** [comment]
Hi @pdlug,

I've been playing around with `hono` in one of my applications and what I ended up doing was something along the lines of:

1. Define the user as optional
```ts
declare module "hono" {
  interface ContextVariableMap {
    user?: User;
  }
}
```

2. Have a custom helper to get the user that takes in the context variable, and throw an error is not authorised
```ts
function getUser(c: Context<any, any, {}>): User {
  const { user } = c.var
  if (!user) { 
    throw new HTTPException(401, { message: 'Not signed in' })
  }
  return c.get('user')
}
```

see https://hono.dev/docs/api/exception

This way you can ensure the code executed beneath the `getUser(c)` will always have the user otherwise they will get a `401` error status.

Hope that helps!

**Comment by karlbateman** [comment]
I've been struggling with this exact same usecase too. I have tried structuring my custom authentication middleware to match `hono/request-id` as this middleware seems to conditionally add the `c.get("requestId")` depending on if it's applied to the routes. For some reason, I cannot get the same behaviour no matter what I try.

@nmcdaines I've ended up using your approach but would love for someone to shed light on how to get this working in the same way the included middlewares conditionally apply typesafe values to the context.

**Comment by maou-shonen** [comment]
Please refer to my reply in another discussion.
https://github.com/orgs/honojs/discussions/3499#discussioncomment-10926890

**Comment by nilobarp** [comment]
An example type aware auth middleware
```
import type { MiddlewareHandler } from "hono";
import { createMiddleware } from "hono/factory";

type Env = {
  Variables: {
    user: {
      id: string;
      username: string;
    };
  };
};

export const useAuth = (): MiddlewareHandler<Env> => {
  return createMiddleware<Env>(async (c, next) => {
    const token = c.req.header("Authorization")?.split("Bearer ")[1];

    if (token === "valid_token") {
      c.set("user", { id: "01K151NXGBEXED3K7CCTXK739X", username: "someuser" });
    } else {
      return c.json({ status: "unauthorized", message: "Invalid token" }, 401);
    }

    await next();
  });
};
```

<img width="412" height="239" alt="image" src="https://github.com/user-attachments/assets/788eb644-5e81-4d40-9afc-1d21238208dc" />

When the middleware isn't applied
<img width="1114" height="244" alt="image" src="https://github.com/user-attachments/assets/d1c680d7-ed9f-42eb-8093-d81beffaa675" />

Caveat:
This solution is only useful when the middleware is applied directly on the route.
